text/0000-min-rust-version.md

```diff
@@ -0,0 +1,179 @@
+tree, but will not be built are excluded from this check (e.g. target-dependent
+or optional crates).
+
+`rust` field should respect the following minimal requirements:
+- Value should be a version in semver format **without** range operators. Note
+that "1.50" is a valid value and implies "1.50.0".
+- Version can not be bigger than a current stable toolchain (it will be checked
+by crates.io during crate upload).
+- Version can not be smaller than 1.27 (version in which  `package.rust` field
+became a warning instead of an error).
+- Version can not be smaller than release version of a used edition, i.e.
+combination of `rust = "1.27"` and `edition = "2018"` is an invalid one.
+
+# Future work and extensions
+[future-work]: #future-work
+
+## Influencing version resolution
+
+The value of `rust` field (explicit or automatically selected by `cargo`) will
+be used to select appropriate dependency versions.
+
+For example, let's imagine that your crate depends on crate `foo` with 10 published
+versions from `0.1.0` to `0.1.9`, in versions from `0.1.0` to `0.1.5` `rust`
+field in the `Cargo.toml` sent to crates.io equals to "1.30" and for others to
+"1.40". Now if you'll build your project with e.g. Rust 1.33, `cargo` will select
+`foo v0.1.5`. `foo v0.1.9` will be selected only if you'll build your project with
+Rust 1.40 or higher. But if you'll try to build your project with Rust 1.29 cargo
+will issue an error.
+
+`rust` field value will be checked as well. During crate build `cargo` will check
+if all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow us to close the long-standing
+debate regarding whether MSRV bump is a breaking change or not and will allow
+crate authors to feel less restrictive about bumping their crate's MSRV. (though
+it may be a useful convention for post-1.0 crates to bump minor version on MSRV
+bump to allow publishing backports which fix serious issues using patch version)
+
+Note that described MSRV constraints and checks for dependency versions resolution
+can be disabled with the `--no-msrv-check` option.
+
+## Checking MSRV during publishing
+
+`cargo publish` will check that upload is done with a toolchain version specified
+in the `rust` field. If toolchain version is different, `cargo` will refuse to
+upload the crate. It will be a failsafe to prevent uses of incorrect `rust` values
+due to unintended MSRV bumps. This check can be disabled by using the existing
+`--no-verify` option.
+
+## Making `rust` field mandatory
+
+In future (probably in a next edition) we could make `rust` field mandatory for
+a newly uploaded crates. MSRV for older crates will be determined by the `edition`
+field. In other words `edition = "2018"` will imply `rust = "1.31"` and
+`edition = "2015"` will imply `rust = "1.0"`.
```

Comment:
Implying Rust 1.0 seems mostly incompatible with this statement from earlier in the document:

> Version can not be smaller than 1.27 (version in which  `package.rust` field became a warning instead of an error).

Reply:
The cited text is one of the requirements for the `rust` field value if it's provided explicitly, while the `edition` text is about how MSRV should be inferred if explicit value is not provided. I guess a better wording will be something like this:
>In other words `edition = "2018"` will imply that MSRV for the crate is equal to 1.31 and `edition = "2015"` will imply that MSRV is equal to 1.0.